### PR TITLE
add form edit validations

### PR DESCRIPTION
**Pull request by jokeewu**
just for complex form validate. 
1.add not required validate, use like '!email'
2.readonly field doesn't validate when form edit
3.when form edit new value equals old do nothing at form edit
4.fixed async validate invalid when submit

### CODE REVIEW

**Comment by huei90** [comment]
Thanks @jokeewu 

seems you are using the old code, do you mind update the latest one and submit again?

then I can review your change.

:beers:

File: src/validator.directive.js [hunk 1]
```diff
@@ -201,6 +261,15 @@
...
       require: 'ngModel',
+      scope: {
+        model: '=ngModel',
+        initialValidity: '=initialValidity',
+        validCallback: '&',
+        invalidCallback: '&',
+        messageId: '@',
+        editReadonly: '=editReadonly',
+        originalData: '=originalData'
+      },
```
Comment (huei90) [inline src/validator.directive.js]: We have put everything under link which allow us to isolate the scope when using multiple directive.

**Comment by huei90** [comment]
Do you mind explain more about the point 1,3,4?

**Comment by jokeewu** [comment]
Case 1:  We just validate when input field is not empty, if user do nothing validate is valid(passed).
Case 3:  When edit a form do nothing for that haven't modified field.(Current value compared with 'originalData', if equal do nothing, or validate it)
Case 4:  To ensure all async validate of form passed, then submit form.  (When a form has many async validate, issue happened when submit form!! you can test it.)

**Comment by huei90** [comment]
Case1,3: What about the case `attribute` has changed? or it is async check?

Case 4: We do support `$q` to handle multiple async request

<hr/>

不好意思，因为您一次提交了多个功能修改，在两个 Commit 中我有点难理解所有的情况，
能否将个别功能拆开来，分别 Commit，方便读取和理解。

抱歉，希望您能够体谅。

File: src/validator.directive.js [hunk 2]
```diff
@@ -18,7 +18,7 @@
      * @param ctrl
      * @returns {}
      */
-    var validFunc = function(element, validMessage, validation, scope, ctrl, attrs) {
+    var validFunc = function(element, validMessage, validation, scope, ctrl, validations, hasValidation) {
```
Comment (huei90) [inline src/validator.directive.js]: Changing the parameters is not recommended.

**Comment by huei90** [comment]
@jokeewu I will do the task if you don't have any update in these few days.

Hope this is fine.

**Pull request closed without merging**